crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs

```diff
@@ -121,78 +136,91 @@ pub mod internal {
-                let pid = sysinfo::get_current_pid().expect("Failed to get current process ID");
-                sys.refresh_processes(sysinfo::ProcessesToUpdate::Some(&[pid]), true);
-
-                sys.refresh_cpu_specifics(CpuRefreshKind::nothing().with_cpu_usage());
-                sys.refresh_memory();
-                let system_cpu_usage = sys.global_cpu_usage().into();
-                let total_mem = sys.total_memory() as f64;
-                let used_mem = sys.used_memory() as f64;
-                let system_mem_usage = used_mem / total_mem * 100.0;
-
-                let process_mem_usage = sys
-                    .process(pid)
-                    .map(|p| p.memory() as f64 * BYTES_TO_GIB)
-                    .unwrap_or(0.0);
-
-                let process_cpu_usage = sys
-                    .process(pid)
-                    .map(|p| p.cpu_usage() as f64 / sys.cpus().len() as f64)
-                    .unwrap_or(0.0);
-
-                SysinfoRefreshData {
-                    system_cpu_usage,
-                    system_mem_usage,
-                    process_cpu_usage,
-                    process_mem_usage,
-                }
-            });
-            tasks.tasks.push(task);
-            *last_refresh = Instant::now();
+    struct DiagnosticTask {
+        system: System,
+        last_refresh: Instant,
+        sender: Sender<SysinfoRefreshData>,
+        waker: Arc<AtomicWaker>,
+    }
+
+    impl DiagnosticTask {
+        fn new(sender: Sender<SysinfoRefreshData>) -> Self {
+            Self {
+                system: System::new_with_specifics(
+                    RefreshKind::nothing()
+                        .with_cpu(CpuRefreshKind::nothing().with_cpu_usage())
+                        .with_memory(MemoryRefreshKind::everything()),
+                ),
+                last_refresh: Instant::now() - sysinfo::MINIMUM_CPU_UPDATE_INTERVAL,
+                sender,
+                waker: Arc::default(),
+            }
         }
     }
 
-    fn read_diagnostic_tasks(mut diagnostics: Diagnostics, mut tasks: ResMut<SysinfoTasks>) {
-        tasks.tasks.retain_mut(|task| {
-            let Some(data) = block_on(poll_once(task)) else {
-                return true;
-            };
+    impl Future for DiagnosticTask {
+        type Output = ();
+
+        fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            self.waker.register(cx.waker());
+
+            if self.last_refresh.elapsed() > sysinfo::MINIMUM_CPU_UPDATE_INTERVAL {
+                self.last_refresh = Instant::now();
+
+                let sysinfo_refresh_data = SysinfoRefreshData::new(&mut self.system);
+                self.sender.send(sysinfo_refresh_data).unwrap();
+            }
 
+            Poll::Pending
+        }
+    }
+
+    fn wake_diagnostic_task(task: ResMut<SysinfoTask>) {
+        if let Some(waker) = task.waker.take() {
+            waker.wake();
+        }
```

Comment:
```suggestion
        task.waker.wake();
```

Alternatively just replace calls to this function with exactly that.